Rando: Enhanced Magic Arrows

Sarge-117 opened the pull request:
This feature, when enabled, will allow Ice Arrows to melt red ice and Light Arrows to activate sun switches.  It also allows Ice Arrows to destroy the mud walls in Dodongo's Cavern, thus mimicking blue fire's properties.  Inspired by the "Blue Fire Arrows" option that is gaining popularity in N64 rando.

Footage:
https://cdn.discordapp.com/attachments/988962246585634856/1011068145118740530/bluefirearrows_concept.mp4
https://cdn.discordapp.com/attachments/988962246585634856/1013173816081514657/lightarrow.mp4

~~**Note** This does not currently change rando's logic in any way.  Logic will need some changes to account for these.~~
~~**Note 2** This is currently set up to be a feature for rando only.~~
~~**Note** Still needs its home in the rando settings menu (#1321) and a way to be locked in while playing on an affected seed~~

Brief implementation explanation:
Overall the way this works is by adjusting colliders on the red ice actor (z_bg_ice_shelter.c), mud wall actor (z_bg_breakwall.c) and sun switch actor (z_obj_lightswitch.c) such that they react to the magic arrows.

For red ice: One of its two cylinder colliders is replaced by a new one which allows for incoming player damage (AC_TYPE_PLAYER).  Then both of its cylinders are checked for a hit by Ice Arrows (which are a child of normal arrows).

For the mud walls: ~~A new quad collider is added, overlapping its existing one.  This one only checks for ice arrows.~~ Its quad collider is replaced by one which allows for incoming Ice Arrow hits, and then the function that checks for hits from bombs also checks for a hits from Ice Arrows.

For the sun switches: Its sphere collider is replaced by one that allows for light arrow hits.  Then some additional logic is added to the switches to remember if/when it's been activated by a light arrow, since some sun switches will turn off when they are no longer in the sunlight.  

When activated by a light arrow, the flag for the switch will be cleared on room exit.  This makes the switch "temporary" when they're activated by a light arrow instead of by sunlight.  The reason for this is because of the rock in the wall on child side spirit - its existence is tied to the switch flag.  If the flag is not cleared, then that rock would disappear by itself when that switch was hit by a light arrow.  

The flag is **not cleared** for the switch in the top floor that controls the lowering chain platform - that one will stay on no matter what to prevent the platform from going back up on its own.  

In the case of being hit by sunlight **after already being activated** by a light arrow, then the switches will behave as if they were originally activated by sunlight, and the flag will not be cleared.

Sarge-117 committed:
Implement Enhanced Magic Arrows

Currently as a rando-only enhancement until it can be accounted for in logic

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Comment clarity

Sarge-117 committed:
Add rando enhancement checkbox

Sarge-117 committed:
Change bool name

To make jenkins happy (I hope)

aMannus commented:
We've talked about this on discord for a little bit already, but I'd like to state it here too. While this is very much on the edge of something that would be considered a mod, I'd still say this should be a base game enhancement (which then rando can hook into).

On one hand, we should have this in rando because it's interesting and also a thing in other randos, so it's also good for parity. On the other hand, it would feel weird not including vanilla players to let them play around with this. I'd love to hear other's opinion on this, but my vote is definitely for adding it to the vanila game enhancements.

Also, slightly related, I'm not entirely sure if we should split them up in 2 seperate options or keep them combined. I'm personally leaning towards splitting them up. If we do end up adding it to the vanilla enhancements and also want to split them up, it might be nice to combine this with https://github.com/HarbourMasters/Shipwright/pull/1022 into a single flyout under "Gameplay" called "Arrows" or something like that.

Sarge-117 commented:
I would agree with letting vanilla players in on this enhancement, as long as other devs agree!  I think it would generally only affect Ganon's castle, since you probably wouldn't normally be in Ice Cavern with Ice Arrows or in Spirit Temple with Light Arrows.

For now though, maybe we let this feature "mature" in rando.  I.E. give it some more testing, get the logic working, etc.  Then if everybody's happy we let it migrate to the base game.

PurpleHato commented:
> I would agree with letting vanilla players in on this enhancement, as long as other devs agree! I think it would generally only affect Ganon's castle, since you probably wouldn't normally be in Ice Cavern with Ice Arrows or in Spirit Temple with Light Arrows.
> 
> For now though, maybe we let this feature "mature" in rando. I.E. give it some more testing, get the logic working, etc. Then if everybody's happy we let it migrate to the base game.

I'm not against to have them in vanilla but since it's initially a rando thingy, I guess the place ofr it needs to be under RANDO ENHANCEMENTS (which is what you did) for clarity since it's where it's initially coming from and where poeple who knows rando will search for it, so all good to me :)

aMannus commented:
> I'm not against to have them in vanilla but since it's initially a rando thingy, I guess the place ofr it needs to be under RANDO ENHANCEMENTS (which is what you did) for clarity since it's where it's initially coming from and where poeple who knows rando will search for it, so all good to me :)

Nah, it would live inside the normal enhancements, and then have an option in the randomizer settings menu to enable it for rando (and have logic be adjusted when it's enabled). That in turn would force the "normal" enhancement to be enabled whenever a rando savefile is loaded with that option enabled.

leggettc18 commented:
Starting out in Rando Enhancements is fine I think, in the interest of not blocking a merge behind this debate. Shouldn't be hard to move it later. The question of whether this should be merged at all is, I think, settled by the fact that we need it to be possible for rando parity (the ice arrows as blue fire at least). As for Light arrows lighting up sun switches, I'm not sure if that's in other randos or not, but if not it does give us a chance to pioneer it!

Sarge-117 commented:
Thank you for the support/feedback @aMannus @PurpleHato @leggettc18!  If any of you are willing to look at hooking this into the rando seed and the logic, that would be much appreciated, since I'm not familiar with the workings of that part.

aMannus commented:
I'm ok with doing the rando part of it, but I'd prefer waiting for https://github.com/HarbourMasters/Shipwright/pull/1321 to be done so I can put it in the proper spot immediately.

Edit: To be clear, I'm ok with this being merged in as it is and move the option into the randomizer settings menu in the future in another PR.

briaguya0 commented:
couple things, first just wanted to note where blue fire arrows exist in 64rando https://ootrandomizer.com/generatorDev?version=devR_

![image](https://user-images.githubusercontent.com/70942617/187474340-2366e008-00ca-42fd-8309-0827c9185efa.png)

second, unless the seed generation logic is updated to account for this, it is not a rando enhancement. it's just a mod. i feel like merging this into rando-next would be "sneaking" it into the next release, and it really should get some other eyes on it.

third, i don't like that light arrows working as sunlight is included. i feel like that should be a separate enhancement as it will require completely separate logic updates

Sarge-117 committed:
Split into two enhancements

"Blue Fire Arrows" and "Sunlight Arrows"

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 commented:
Okay, I have split this into two - "Blue Fire Arrows" and "Sunlight Arrows".  They still live in their temporary home as basic CVars since I believe @aMannus has a specific place he wants to put them once #1321 is ready.

> second, unless the seed generation logic is updated to account for this, it is not a rando enhancement. it's just a mod. i feel like merging this into rando-next would be "sneaking" it into the next release, and it really should get some other eyes on it.

Agreed on this.  I will be relying on help from others for the logic and seed generation, since I have no experience there.  In its current form, this PR is essentially just for testing the arrows' function.

garrettjoecox commented:
I can take a whack at it when I get home from work, but it would probably be good for more people to get familiar with the rando logic.

A new helper is needed here that encapsulates HasBlueFire, CanUseMagic, and HasIceArrows https://github.com/HarbourMasters/Shipwright/blob/develop-zhora/soh/soh/Enhancements/randomizer/3drando/logic.cpp

then for each of the actual checks you’ll look in https://github.com/HarbourMasters/Shipwright/blob/develop-zhora/soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ice_cavern.cpp and any other places necessary, and replace the hasBlueFire with your new helper

(typed this from phone sorry for formatting)

leggettc18 commented:
The logic also has to account for the setting being turned on, and I don't even think 3DS has this setting so we'd have to add one. Shouldn't be too hard to do, just adding it to the notes.

I think it would be better to extend `HasBlueFire` to include the setting being on, `CanUseMagic` and `HasIceArrows` rather than making a new helper. That way instead of adding this new helper everywhere, it will automatically apply in logic everywhere blue fire does while only changing one line. 

Light arrows on sun switches will definitely require brand new logic though. It can work in some situations that Mirror shield alone can't. For instance, the Spirit Trial, if you have Light Arrows and Magic, you can clear it with this setting, but without it you require Dins or Fire Arrows and Magic along with Mirror Shield. I think you also need Bombs to get the light beam to the platform to the boss in Spirit Temple, but with Light Arrows you wouldn't need those.

stratomaster64 commented:
> Light arrows on sun switches will definitely require brand new logic though. It can work in some situations that Mirror shield alone can't. For instance, the Spirit Trial, if you have Light Arrows and Magic, you can clear it with this setting, but without it you require Dins or Fire Arrows and Magic along with Mirror Shield. I think you also need Bombs to get the light beam to the platform to the boss in Spirit Temple, but with Light Arrows you wouldn't need those.

We may need to be careful since although sunlight arrows can bypass *most* mirror shield checks, it's still a hard requirement for Twinrova. Other than that, it'd be a pretty nifty thing, especially as rando progress gets to the point that we can mix in MQ dungeons.

Sarge-117 commented:
Yes, light arrows would let you lower the chain platform, but you still need mirror shield to open the face of the statue and for Twinrova.

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
First pass on logic

Sarge-117 commented:
Did a pass on the rando logic for these.  Hopefully I didn't overlook too much!  The Cvars are still under the rando enhancements menu for now.

briaguya0 commented:
The rando logic is looking good! Now the only missing puzzle piece is having cvars tied to rando settings/saves. @aMannus any thoughts on this one?

aMannus commented:
https://github.com/HarbourMasters/Shipwright/pull/1321 has a spot for them inside the randomizer settings menu (which is ready for review/feedback btw winkwink). Once it's in there, it can be handled just like any other randomizer setting.

briaguya0 commented:
> #1321 has a spot for them inside the randomizer settings menu (which is ready for review/feedback btw winkwink). Once it's in there, it can be handled just like any other randomizer setting.

so are you suggesting just updating this PR to use the RandomizerSetting instead of the CVar everywhere? I thought the idea would be to have the RandomizerSetting force the CVar when set as a part of a seed, but still allow vanilla or rando players without fancy arrows in logic to use it out of logic

aMannus commented:
> > #1321 has a spot for them inside the randomizer settings menu (which is ready for review/feedback btw winkwink). Once it's in there, it can be handled just like any other randomizer setting.
> 
> so are you suggesting just updating this PR to use the RandomizerSetting instead of the CVar everywhere? I thought the idea would be to have the RandomizerSetting force the CVar when set as a part of a seed, but still allow vanilla or rando players without fancy arrows in logic to use it out of logic

It should live in the general SoH enhancements so we can then force that setting to be on when a rando savefile is loaded with the setting enabled (through the randomizer settings menu). I was under the impression this PR was going to get merged in before it was also added to the regular enhancements, but I might've misunderstood that part then.

If we want to do that instead, it should be PR'd to Zhora first, and then picked up in rando-next afterwards.

It doesn't make much sense to me having the magical arrows live in the rando enhancements menu, and then also in the randomizer settings menu.

briaguya0 commented:
> It doesn't make much sense to me having the magical arrows live in the rando enhancements menu, and then also in the randomizer settings menu.

that's my thought too

so @Sarge-117 you'll want to cherry pick together a new branch off zhora that stops at https://github.com/HarbourMasters/Shipwright/pull/1335/commits/8e2f035278734cd71a993de5149623af50532d24, make a PR to zhora with all of that, get that reviewed (be sure to mention this PR in that PR so it doesn't get written off as a "this should be a mod" thing) and merged, then i'll do a zhora -> rando-next and we can do a final pass on this PR with all the logic

Sarge-117 commented:
Sounds good!

Just to be clear: we'll want the conditionals on the red ice/mudwall/sun switch actors to only check for if(cvar), not for if(n64dd && cvar), correct?  I will make that change first and remove the checkboxes from the rando enhancements flyout.

aMannus commented:
> Sounds good!
> 
> Just to be clear: we'll want the conditionals on the red ice/mudwall/sun switch actors to only check for if(cvar), not for if(n64dd && cvar), correct? I will make that change first and remove the checkboxes from the rando enhancements flyout.

For the zhora one, just the CVar.

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Check for CVar only

Instead of (Cvar && n64dd)

Sarge-117 committed:
Fix missed comma

Sarge-117 committed:
Move checkboxes to normal enhancement menu

Also comment clarity + capitalization

Sarge-117 committed:
Add magic arrow enhancements, non-rando

Changes to red ice actor, sun switch actor, and mudwall actor to adjust their collision

Sarge-117 committed:
Update tooltip

Sarge-117 commented:
Created PR #1376 for the non-rando parts of this, with options placed under Enhancements > Gameplay

Sarge-117 committed:
Update tooltip

Sarge-117 committed:
Update logic for King Dodongo's room

Explicitly check for BlueFireAccess && HasBottle here, since using the Ice Arrows on the bombable floor is tricky due to its collision being very tiny

leggettc18 commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_dodongos_cavern.cpp:
```diff
@@ -235,7 +235,7 @@ void AreaTable_Init_DodongosCavern() {
                   //Events
                   EventAccess(&DodongosCavernClear, {[]{return DodongosCavernClear || (Here(DODONGOS_CAVERN_BOSS_ROOM, []{return HasExplosives || (CanUse(MEGATON_HAMMER) && CanShield);}) && (Bombs || GoronBracelet) && (IsAdult || Sticks || KokiriSword));},
-                                         /*Glitched*/[]{return Here(DODONGOS_CAVERN_BOSS_ROOM, []{return HasExplosives || (CanUse(MEGATON_HAMMER) && CanShield) || (GlitchBlueFireWall && BlueFire);}) && (HasExplosives || GoronBracelet) && (IsAdult || Sticks || KokiriSword || CanUse(MEGATON_HAMMER));}}),
+                                         /*Glitched*/[]{return Here(DODONGOS_CAVERN_BOSS_ROOM, []{return HasExplosives || (CanUse(MEGATON_HAMMER) && CanShield) || (GlitchBlueFireWall && BlueFire && HasBottle && BlueFireAccess);}) && (HasExplosives || GoronBracelet) && (IsAdult || Sticks || KokiriSword || CanUse(MEGATON_HAMMER));}}),
```
Just making sure I'm following the logic, this is here because the Blue Fire Arrows don't work for the bombable floor above KD right? But the bottle of blue fire does still work? Is it that way in N64 rando as well?

leggettc18 replied:
Looks like you explained in your commit message. Which way I swing on this one is gonna depend on what n64 rando does. If they also have their logic set up this way then we can keep that as is. If not, I'd rather us take the time to figure out how to make the collision on the bombable floor work better.

leggettc18 replied:
Just checked, if I'm reading their Glitched World JSON files right, they don't consider having Blue Fire as being able to reach King Dodongo, so we're all good! Disregard these comments.

aMannus commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_dodongos_cavern.cpp:
```diff
@@ -235,7 +235,7 @@ void AreaTable_Init_DodongosCavern() {
                   //Events
                   EventAccess(&DodongosCavernClear, {[]{return DodongosCavernClear || (Here(DODONGOS_CAVERN_BOSS_ROOM, []{return HasExplosives || (CanUse(MEGATON_HAMMER) && CanShield);}) && (Bombs || GoronBracelet) && (IsAdult || Sticks || KokiriSword));},
-                                         /*Glitched*/[]{return Here(DODONGOS_CAVERN_BOSS_ROOM, []{return HasExplosives || (CanUse(MEGATON_HAMMER) && CanShield) || (GlitchBlueFireWall && BlueFire);}) && (HasExplosives || GoronBracelet) && (IsAdult || Sticks || KokiriSword || CanUse(MEGATON_HAMMER));}}),
+                                         /*Glitched*/[]{return Here(DODONGOS_CAVERN_BOSS_ROOM, []{return HasExplosives || (CanUse(MEGATON_HAMMER) && CanShield) || (GlitchBlueFireWall && BlueFire && HasBottle && BlueFireAccess);}) && (HasExplosives || GoronBracelet) && (IsAdult || Sticks || KokiriSword || CanUse(MEGATON_HAMMER));}}),
```
Why is this neccesary when the conditionals you added here already exist in `BlueFire`?

leggettc18 replied:
They are overriding blue fire arrows being in logic here, since breaking the floor over King Dodongo's room is not in logic with blue fire arrows in N64 rando. Actually it doesn't seem to be in logic with regular Blue Fire over there either, but the other pieces of this were already here in 3DS rando. Apparently the hitbox of that floor is difficult to hit with the ice arrows.

aMannus replied:
That might be the intent, but that's not what's happening here. Then BlueFire in here should be removed.

Edit: actually, end result might be correct, but BlueFire is still redundant.

leggettc18 replied:
I initially did say that it was fine since N64 also didn't have Blue Fire Arrows as logically able to enter, but since they also don't have blue fire in logic at all for getting to KD I do want to revisit that statement. I'm inclined to say that it's OK if the hitbox for the floor is a bit tricky with Blue Fire Arrows since it would only be required in glitched logic anyway.

aMannus replied:
I'm fine with the logic being different for this, I'm saying `BlueFire` is redundant when `HasBottle && BlueFireAccess` is already in the conditional.

As a side note, this part of the logic isn't even going to be used until we add glitches & tricks. GlitchBlueFireWall is a trick on 3DS, only red ice is in logic without it enabled. But it's good to have it ready either way.

Sarge-117 replied:
Oops, right.  I guess I don't need the `BlueFire` part in there.

And yeah, the floor's hitbox there basically a tiny, nearly pixel-perfect spot in the dead centre of that floor.  It works fine for AOE stuff like bombs/hammer/regular blue fire, but it's hard to hit with arrows.  If you didn't have knowledge of this particular hitbox, you'd probably end up quite confused (if your seed expected you to use the Ice Arrows here).

Sarge-117 committed:
Remove redundancy in KD room

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Update logic Cvar names

aMannus commented:
Might be good to cross reference the logic with https://github.com/gamestabled/OoT3D_Randomizer/pull/543.

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge remote-tracking branch 'upstream/develop-zhora' into zhora-magic-arrows

Sarge-117 committed:
Fix mistake from merge

oops

Sarge-117 committed:
Restore checkboxes to enhancements menu

These got lost in the merge

Sarge-117 committed:
Update location access logic

Including MQ locations in Spirit and GC now

Sarge-117 committed:
Simplify mudwall collision check

There was no need to have a second collider specifically for Ice Arrow hits

Sarge-117 committed:
Simplify mudwall collision check

Sarge-117 committed:
Restore checkboxes in menu

Accidentally lost these during merge

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge branch 'rando-next' into rando-enhanced-magic-arrows

Sarge-117 commented:
Update Sept 3:
Updated location access logic for Spirit Temple and Ganon's Castle (now includes all MQ checks as well) - credit to HylianFreddy from https://github.com/gamestabled/OoT3D_Randomizer/pull/543
Simplified the mudwall collision check for Ice Arrows
Got this up to date with latest rando-next changes

Sarge-117 committed:
Update logic Cvar name

aMannus committed:
Incorporate magic arrows in rando settings

Sarge-117 committed:
Merge branch 'rando-enhanced-magic-arrows' into rando-magical-arrows

Sarge-117 committed:
Merge pull request #11 from aMannus/rando-magical-arrows

Incorporate magic arrows in rando settings

Sarge-117 committed:
Update tooltip

To inform the player that they might have to reload the room if they're enabling this for the first time.

Sarge-117 committed:
Update tooltip

Sarge-117 committed:
Add line break in tooltip

Sarge-117 committed:
Tooltip wording + line break

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge remote-tracking branch 'upstream/develop-zhora' into zhora-magic-arrows

Sarge-117 committed:
Clarify var names and comments

Also preserve chain platform cutscene in spirit based on Link's position

Sarge-117 committed:
Remove !=0 from cvar check

Sarge-117 committed:
Clarify var names and comments

Sarge-117 committed:
Merge branch 'zhora-magic-arrows' into rando-enhanced-magic-arrows

Sarge-117 committed:
Fix merge mistake

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Tidy up chain platform cutscene check

Sarge-117 committed:
Fix merge error

Didn't mean to have Zhora changes in here yet

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows

Sarge-117 committed:
Fix merge error

Sarge-117 committed:
Merge remote-tracking branch 'upstream/rando-next' into rando-enhanced-magic-arrows
